core/src/test/java/com/linecorp/armeria/common/stream/StreamMessageInputStreamTest.java

```diff
@@ -0,0 +1,317 @@
+        int read;
+        while ((read = inputStream.read()) != -1) {
+            result.writeByte(read);
+        }
+
+        final int readableBytes = result.readableBytes();
+        final byte[] actual = new byte[readableBytes];
+        for (int i = 0; i < readableBytes; i++) {
+            actual[i] = result.readByte();
+        }
+        result.release();
+        assertThat(actual).isEqualTo(expected);
+        assertThat(inputStream.available()).isZero();
+    }
+
+    @Test
+    void readWithOffset() throws Exception {
+        final StreamMessage<byte[]> streamMessage = StreamMessage.of(new byte[] {1, 2, 3, 4, 5});
+        final InputStream inputStream = streamMessage.toInputStream(HttpData::wrap);
+
+        final byte[] result = new byte[5];
+        final int len = inputStream.read(result, 1, 3);
+
+        assertThat(len).isEqualTo(3);
+        assertThat(result).isEqualTo(new byte[] { 0, 1, 2, 3, 0});
+        assertThat(inputStream.available()).isEqualTo(2);
+    }
+
+    @Test
+    void close() throws Exception {
+        final Publisher<Integer> publisher = Flux.range(1, 10);
+        final StreamMessage<Integer> streamMessage = new PublisherBasedStreamMessage<>(publisher);
+        final InputStream inputStream = streamMessage
+                .toInputStream(x -> HttpData.wrap(x.toString().getBytes()));
+        final byte[] expected = ImmutableList.of(1, 2, 3, 4, 5)
+                                             .stream()
+                                             .map(x -> x.toString().getBytes())
+                                             .reduce(Bytes::concat).get();
+
+        final ByteBuf result = Unpooled.buffer();
+        int read;
+        while ((read = inputStream.read()) != -1) {
+            if (result.readableBytes() == 5) {
+                inputStream.close();
+                break;
+            }
+            result.writeByte(read);
+        }
+
+        final int readableBytes = result.readableBytes();
+        final byte[] actual = new byte[readableBytes];
+        for (int i = 0; i < readableBytes; i++) {
+            actual[i] = result.readByte();
+        }
+        result.release();
+        assertThat(actual).isEqualTo(expected);
+        assertDoesNotThrow(inputStream::close);
+        assertThatThrownBy(inputStream::read).isInstanceOf(IOException.class)
+                                             .hasMessage("Stream closed");
+        assertThatThrownBy(inputStream::available).isInstanceOf(IOException.class)
+                                                  .hasMessage("Stream closed");
+    }
+
+    @Test
+    void closeBeforeRead() throws IOException {
+        final StreamMessage<String> streamMessage = StreamMessage.of("foo", "bar", "baz");
+        final InputStream inputStream = streamMessage.toInputStream(x -> HttpData.wrap(x.getBytes()));
+        assertThat(inputStream.available()).isZero();
+
+        assertDoesNotThrow(inputStream::close);
```

Comment:
Could we add an assertion that make sure `streamMessage` is closed by `InputStream`?
```
await().untilAsserted(() -> {
    assertThat(streamMessage.whenComplete()).isDone();
});
```

Reply:
oh this is really good point! thanks 🙇 
✅  updated.